torch/functional.py

```diff
@@ -1,3 +1,5 @@
+from typing import Tuple, Optional
```

Comment:
So adding this line, which is definitely correct, is the reason for the CI failure that can be reproduced with:
```
$ python -mmypy --follow-imports silent --check-untyped-defs test/generated_type_hints_smoketest.py 
test/generated_type_hints_smoketest.py:903: error: Argument 1 to "bmm" has incompatible type "Optional[Tensor]"; expected "Tensor"  [arg-type]
test/generated_type_hints_smoketest.py:903: error: Argument 2 to "bmm" has incompatible type "Optional[Tensor]"; expected "Tensor"  [arg-type]
test/generated_type_hints_smoketest.py:911: error: Argument 1 to "bmm" has incompatible type "Optional[Tensor]"; expected "Tensor"  [arg-type]
test/generated_type_hints_smoketest.py:911: error: Argument 2 to "bmm" has incompatible type "Optional[Tensor]"; expected "Tensor"  [arg-type]
```

The annotation of `lu_unpack` is:
```
def lu_unpack(LU_data, LU_pivots, unpack_data=True, unpack_pivots=True):
    # type: (Tensor, Tensor, bool, bool) ->  (Tuple[Optional[Tensor], Optional[Tensor], Optional[Tensor]])
```
which seems correct (but was ignored before because of the missing import I think), because the implementation can return `None` for any of the outputs, depending on the `bool` inputs.

The failure can be fixed by adding `--no-strict-optional`. Which may be the best option here, none of the alternatives in [the MyPy FAQ on this](https://mypy.readthedocs.io/en/stable/kinds_of_types.html#strict-optional) seem appealing.

This really is an issue with how `lu_unpack` is written; the bool flags are not very nice from either a typing or API design perspective. Splitting the function into two separate ones could be considered.

Reply:
Oh yes, we need some sort of dependent types to type this interface correctly. Fortunately, mypy does support this! See https://mypy.readthedocs.io/en/stable/generics.html#type-variable-value-restriction

I don't mind temporarily adding this so we can land this diff though.

Reply:
I don't immediately see how that value restriction helps. The output type here is determined not by the _types_ of the inputs, but by the _values_ of the bool keywords. I don't think you can determine those values with static analysis.

Reply:
Maybe for this particular test it doesn't help. But the idea behind value restriction is if a user writes something like `f(x=True)` (where the boolean value is manifestly obvious), mypy will pick the appropriate overload